greenbone/scap/generic_cli/producer/db.py

```diff
@@ -0,0 +1,317 @@
+                f"schema: {database_schema}, user: {database_user}"
+            )
+
+        self._manager = self._create_manager()
+
+    @abstractmethod
+    def _create_manager(self) -> AsyncContextManager:
+        """
+        Callback creating a new database manager for handling SCAP items.
+
+        Returns: The new database manager.
+        """
+        pass
+
+    @abstractmethod
+    def _convert_db_model(self, db_model: BaseDatabaseModel) -> T:
+        """
+        Callback converting a SCAP database model to a Pontos model.
+
+        Args:
+            db_model: The database model convert
+
+        Returns:
+            The converted model object.
+        """
+        pass
+
+    @abstractmethod
+    async def _db_item_count(self) -> int:
+        """
+        Callback getting the total number of SCAP items in the database.
+
+        Returns:
+            The total number of items
+        """
+        pass
+
+    @abstractmethod
+    def _db_item_iter(self) -> AsyncIterator[BaseDatabaseModel]:
+        """
+        Callback getting an async iterator of database items to process.
+
+        Returns:
+            An async iterator over database items.
+        """
+
+    async def fetch_initial_data(self) -> int:
+        """
+        Ensures any remaining initializations of the data source are done,
+         so it can be queried for chunks of SCAP items by `run_loop`.
+        It must also return the expected total number of items that will be
+         fetched.
+
+        Returns:
+            The expected total number of items.
+        """
+        count = await self._db_item_count()
+        self._console.log(
+            f"{count:,} {self._item_type_plural} available in database"
+        )
+        return count
+
+    async def run_loop(self) -> None:
+        """
+        Run a loop fetching chunks of SCAP items and adding them to the queue.
+
+        The method must also call `set_producer_finished` before returning to signal
+        that no more chunks will be added to the queue.
+
+        It should also create a task for the `progress` object and update it
+        regularly.
+        """
+        task = self._progress.add_task(
+            f"Querying {self._item_type_plural}",
+            total=self._queue.total_items,
+        )
+        count = 0
+
+        try:
+            with Timer() as query_timer:
+                chunk = []
+                async for db_item in self._db_item_iter():
+                    item: T = self._convert_db_model(db_item)
+                    chunk.append(item)
+                    if len(chunk) >= self._queue.chunk_size:
+                        count += len(chunk)
+                        await self._queue.put_chunk(chunk)
+                        chunk = []
+                    self._progress.update(task, completed=count)
+
+                count += len(chunk)
+                if len(chunk):
+                    await self._queue.put_chunk(chunk)
```

Comment:
Wakeup the producer just a last time if it is currently waiting for getting a chunk from the queue.
```suggestion
                await self._queue.put_chunk(chunk)
```